languages/python/reference/exercise-concepts/robot-simulator.md

```diff
@@ -0,0 +1,62 @@
+# Concepts for robot-simulator
+
+[Example Implementation](https://github.com/exercism/python/blob/master/exercises/robot-simulator/example.py)
+
+
+## General
+
+- contstants: Are not enforced by the runtime, but are used via the convention of `<ALL_CAPS_NAMING>` to signal that these values are expected to remaine unchanged.  Perferably, constants are defined at a module level. The example solution uses the ALL_CAPS convention to define NORTH, SOUTH, EAST, and WEST constants. 
+
+- multiple assignment:  Python allows multiple assignment, assiging the items on the left of `=` _in order_ to the values on the right of `=`.  Equivelent to forming tuples of each side and then "unpacking" them.  This exercise solution uses multiple assignment for the contant values NORTH, EAST, SOUTH, WEST.
+
+- `range()` built-in type:  Represents an immutable sequence of numbers (or any object that implements the `__index__` magic method).  Used in the example to represnet the values from zero to 3 as assigned to NORTH, EAST, SOUTH, WEST. 
+
+- classes: The exercise objective is to define a `robot` type.  Tested methods are linked to a `robot` class.
+
+- objects: Creating different instances of the `robot` class with different data representing different starting positions and bearing are tested.
+
+- constructor: Customizing object initalization with actions and persisting data. The example uses a constructor (via `__init__` + `__new__`) to persist a `compass` object and x, y coordinates assigned to instance properties.
+
+- retun values:  Knowing that functions need not have explicit return statements or values but will return `none` if `return` is not specified.  Except for the two `@property` functions, all of the functions in the example omit an explicit `return` statments, and all return `none`.
+
+- self: The example uses this keyword for methods and properties linked to a specific instance of the class.
+
+- namespaces: Knowing to use self.<propertyname> for instance properties and self as first argument to instance methods in a class.  Additionally, the example uses `self.<methodname>()` to call a previously stored function name.
+  
+- instance methods: Tests for this exercises require one or more instance methods that will take in a set of starting coordinates and a bearing and then accept a seiries of instructions that "move" the instance to a new set of coordinates and bearing.
+
+- function decorator: a higer-order function that takes another function as an argument. The "decorating" function extends the behavior of the "decorated" function without explicitly modifying it (i.e. it _wraps_ or _decorates_ it). Called in python by using the `@` symbol and the function name ahead of the function being decorated.  The example uses pythons built-in `property()` as a decorator (`@property`) to return a "compound" read-only instance property made up of two seperate instance properties.
```

Comment:
Spelling: "higer" -> "higher", "seperate" -> "separate". Also not sure we want to get into this, but a decorator can be _any_ callable, not necessarily a function. Classes can be used for decorators, when there's some reason to keep state around.

Reply:
Maybe we separate the two as sub-categories of "decorator"?  For those coming from a functional language, higher-order functions and decorators might feel familiar, but the notion that _classes_ can decorate might be harder to grok.  Conversely, a more OOP-oriented learner might be fine with the class-as-decorator notion, but struggle a bit with the whole function-returns-a-function concept.  Hard for me to think about what the more general language would be to introduce the `decorator` concept though.

Reply:
Maybe just rename this one to "decorator" instead of "function decorator" ... I don''t want to blow too many minds with class-based decorator, but they do get used.